pkg/ebpf/dns_cache.go

```diff
@@ -0,0 +1,204 @@
+package ebpf
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type reverseDNSCache struct {
+	mux  sync.Mutex
+	data map[util.Address]*dnsCacheVal
+	exit chan struct{}
+	ttl  time.Duration
+	size int
+
+	// Telemetry
+	len      int64
+	lookups  int64
+	resolved int64
+}
+
+func newReverseDNSCache(size int, ttl, expirationPeriod time.Duration) *reverseDNSCache {
+	cache := &reverseDNSCache{
+		data: make(map[util.Address]*dnsCacheVal),
+		exit: make(chan struct{}),
+		ttl:  ttl,
+		size: size,
+	}
+
+	ticker := time.NewTicker(expirationPeriod)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				cache.expire()
+			case <-cache.exit:
+				return
+			}
+		}
+	}()
+	return cache
+}
+
+func (c *reverseDNSCache) Add(translation *translation, now time.Time) bool {
+	if translation == nil {
+		return false
+	}
+
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	if len(c.data) >= c.size {
+		return false
+	}
+
+	exp := now.Add(c.ttl).Unix()
+	for addr := range translation.ips {
+		val, ok := c.data[addr]
+		if ok {
+			val.expiration = exp
+			val.merge(translation.name)
+			continue
+		}
+
+		c.data[addr] = &dnsCacheVal{names: []string{translation.name}, expiration: exp}
+	}
+
+	// Update cache length for telemetry purposes
+	atomic.StoreInt64(&c.len, int64(len(c.data)))
+
+	return true
+}
+
+func (c *reverseDNSCache) Get(conns []ConnectionStats, now time.Time) map[util.Address][]string {
+	if len(conns) == 0 {
+		return nil
+	}
+
+	var (
+		ipToNames  = make(map[util.Address][]string)
+		visited    = make(map[util.Address]struct{})
```

Comment:
You don't need `visited` since you can just check if `ipToNames` has been populated already

Reply:
I do. `ipToNames` will only contain the names that were resolved. I also don't want to repeat the lookups for those IPs without names.

Reply:
(In addition to that `visited` also gives us a good telemetry which is total number of `lookups`)

Reply:
I don't know if that's technically # of lookups. It gives you the number of unique lookups which I wouldn't say is the same (or honestly all that useful). It's easy enough to track this with a counter.

We'd probably save a bunch of space if you just store a `nil` slice for those misses and simplify a lot of this code making it more maintainable.

Happy to wait for benchmarks to see # of allocations and time per op though.

Reply:
* I do think having such a metric is important, though I agree we don't need an extra map for that. It's important to know in my opinion how many IPs we have resolved names vs how many don't;
* Having one map allocates `N` entries, where N is the number of unique IPs in all of your `ConnectionStats`. Having 2 maps (`visited` + `ipToNames`) allocates `N` if your resolve ratio is 0% up to `2N` if your resolve ratio is 100%;
* What *I* personally don't like  about what you're suggesting is that we would have maps in the payloads with nil entries. If you think that's OK/better, I'm happy to change the code.

Reply:
> What I personally don't like about what you're suggesting is that we would have maps in the payloads with nil entries. If you think that's OK/better, I'm happy to change the code. In terms of API I don't like the idea of having entries for which we don't have data (names) and that our consumers have to check if an entry exists AND if it's nil (though you might say that doesn't matter since nil is the zero value of a slice)

I think it's totally fine to cache the state that there are no dns entries for an address. Even putting an empty slice is fine though yeah, I was going for nil as the empty slice.

I don't want to focus too much on this, but we need to be pretty conscious of memory usage on the system-probe in general. We can keep it as is for now, but this is probably the first thing I'd change if I needed to do another round of memory improvements...

Reply:
I've refactored the code to reduce the allocations see [this commit](https://gist.github.com/p-lambert/ad111a5bb05ab3a143fde65b24e0b195).

I benchmarked 3 different approaches:
* the one with `visited` map;
* one with a `unresolved` map (the new commit mentioned above);
* one with a single map as you suggested [check this gist](https://gist.github.com/p-lambert/ad111a5bb05ab3a143fde65b24e0b195);

Results:
```
benchcmp visited_map.txt single_map_with_nil_entries.txt
benchmark                                      old ns/op     new ns/op     delta
BenchmarkDNSCacheGet/ResolveRatio-0.000000     42870         64053         +49.41%
BenchmarkDNSCacheGet/ResolveRatio-0.250000     59247         69380         +17.10%
BenchmarkDNSCacheGet/ResolveRatio-0.500000     103074        80299         -22.10%
BenchmarkDNSCacheGet/ResolveRatio-0.750000     109753        86207         -21.45%
BenchmarkDNSCacheGet/ResolveRatio-1.000000     119618        86950         -27.31%

benchmark                                      old allocs     new allocs     delta
BenchmarkDNSCacheGet/ResolveRatio-0.000000     11             13             +18.18%
BenchmarkDNSCacheGet/ResolveRatio-0.250000     57             53             -7.02%
BenchmarkDNSCacheGet/ResolveRatio-0.500000     136            120            -11.76%
BenchmarkDNSCacheGet/ResolveRatio-0.750000     162            165            +1.85%
BenchmarkDNSCacheGet/ResolveRatio-1.000000     224            213            -4.91%

benchmark                                      old bytes     new bytes     delta
BenchmarkDNSCacheGet/ResolveRatio-0.000000     10853         25635         +136.20%
BenchmarkDNSCacheGet/ResolveRatio-0.250000     17140         26258         +53.20%
BenchmarkDNSCacheGet/ResolveRatio-0.500000     37621         27300         -27.43%
BenchmarkDNSCacheGet/ResolveRatio-0.750000     38056         28076         -26.22%
BenchmarkDNSCacheGet/ResolveRatio-1.000000     39629         28826         -27.26%
```

```
benchcmp visited_map.txt unresolved_map.txt
benchmark                                      old ns/op     new ns/op     delta
BenchmarkDNSCacheGet/ResolveRatio-0.000000     42870         42112         -1.77%
BenchmarkDNSCacheGet/ResolveRatio-0.250000     59247         59704         +0.77%
BenchmarkDNSCacheGet/ResolveRatio-0.500000     103074        82830         -19.64%
BenchmarkDNSCacheGet/ResolveRatio-0.750000     109753        89853         -18.13%
BenchmarkDNSCacheGet/ResolveRatio-1.000000     119618        84908         -29.02%

benchmark                                      old allocs     new allocs     delta
BenchmarkDNSCacheGet/ResolveRatio-0.000000     11             11             +0.00%
BenchmarkDNSCacheGet/ResolveRatio-0.250000     57             56             -1.75%
BenchmarkDNSCacheGet/ResolveRatio-0.500000     136            126            -7.35%
BenchmarkDNSCacheGet/ResolveRatio-0.750000     162            168            +3.70%
BenchmarkDNSCacheGet/ResolveRatio-1.000000     224            213            -4.91%

benchmark                                      old bytes     new bytes     delta
BenchmarkDNSCacheGet/ResolveRatio-0.000000     10853         10853         +0.00%
BenchmarkDNSCacheGet/ResolveRatio-0.250000     17140         17008         -0.77%
BenchmarkDNSCacheGet/ResolveRatio-0.500000     37621         31881         -15.26%
BenchmarkDNSCacheGet/ResolveRatio-0.750000     38056         29763         -21.79%
BenchmarkDNSCacheGet/ResolveRatio-1.000000     39629         28831         -27.25%
```

As you can see the new approach has same number of allocation as with the single map, with the advantage that we won't generate sparse maps with nil entries on it.

Reply:
Nice - this looks great, thank you